feature/experimental/leader_failover_test/leader_failover_test.go

```diff
@@ -0,0 +1,371 @@
+
+	ate.Traffic().Start(t, flow)
+	time.Sleep(15 * time.Second)
+	ate.Traffic().Stop(t)
+
+	time.Sleep(time.Minute)
+
+	flowPath := ate.Telemetry().Flow(flow.Name())
+
+	if !wantLoss {
+		if got := flowPath.LossPct().Get(t); got != 0 {
+			t.Errorf("FAIL: LossPct for flow named %s got %g, want 0", flow.Name(), got)
+		} else {
+			t.Logf("LossPct for flow named %s got %g, want 0", flow.Name(), got)
+		}
+	} else {
+		if got := flowPath.LossPct().Get(t); got != 100 {
+			t.Errorf("FAIL: LossPct for flow named %s got %g, want 100", flow.Name(), got)
+		} else {
+			t.Logf("LossPct for flow named %s got %g, want 100", flow.Name(), got)
+		}
+	}
+}
+
+// testArgs holds the objects needed by the test case.
+type testArgs struct {
+	ctx context.Context
+	dut *ondatra.DUTDevice
+	ate *ondatra.ATEDevice
+	top *ondatra.ATETopology
+}
+
+// testIPv4RouteAdd configures an IPV4 Entry through the given gRIBI client
+// and ensures that the entry is active by checking AFT Telemetry and Traffic.
+func testIPv4RouteAdd(ctx context.Context, t *testing.T, args *testArgs, clientA *gribi.Client) {
+
+	// Add an IPv4Entry for 203.0.113.0/24 pointing to ATE port-2 via gRIBI.
+	t.Logf("Add an IPv4Entry for %s pointing to ATE port-2 via clientA", ateDstNetCIDR)
+	clientA.AddNH(t, nhIndex, atePort2.IPv4, instance, fluent.InstalledInRIB)
+	clientA.AddNHG(t, nhgIndex, map[uint64]uint64{nhIndex: 1}, instance, fluent.InstalledInRIB)
+	clientA.AddIPv4(t, ateDstNetCIDR, nhgIndex, instance, "", fluent.InstalledInRIB)
+
+	// Verify the entry for 203.0.113.0/24 is active through AFT Telemetry.
+	t.Logf("Verify through AFT Telemetry that 203.0.113.0/24 is active")
+	ipv4Path := args.dut.Telemetry().NetworkInstance(instance).Afts().Ipv4Entry(ateDstNetCIDR)
+	if got, want := ipv4Path.Prefix().Get(t), ateDstNetCIDR; got != want {
+		t.Errorf("ipv4-entry/state/prefix got %s, want %s", got, want)
+	}
+
+	// Verify by running traffic that the route entry for 203.0.113.0/24 is installed and active.
+	t.Logf("Verify by running traffic that 203.0.113.0/24 is active")
+	srcEndPoint := args.top.Interfaces()[atePort1.Name]
+	dstEndPoint := args.top.Interfaces()[atePort2.Name]
+
+	testTraffic(t, args.ate, args.top, srcEndPoint, dstEndPoint, false)
+}
+
+func TestLeaderFailover(t *testing.T) {
+
+	start := time.Now()
+	dut := ondatra.DUT(t, "dut")
+	ctx := context.Background()
+
+	// Configure the DUT.
+	t.Logf("Configure DUT")
+	configureDUT(t, dut)
+
+	// Configure the ATE.
+	t.Logf("Configure ATE")
+	ate := ondatra.ATE(t, "ate")
+	top := configureATE(t, ate)
+	top.Push(t).StartProtocols(t)
+
+	t.Logf("Time check: %s", time.Since(start))
+
+	args := &testArgs{
+		ctx: ctx,
+		dut: dut,
+		ate: ate,
+		top: top,
+	}
+
+	// Set parameters for gRIBI client clientA.
+	clientA := &gribi.Client{
+		DUT:                  dut,
+		FibACK:               false,
+		Persistence:          false,
+		InitialElectionIDLow: 10,
+	}
+
+	t.Run("SINGLE_PRIMARY/PERSISTENCE=DELETE", func(t *testing.T) {
```

Comment:
The following suggestion may be a little bit prescriptive and experimental, but how do you think about nesting the t.Run similar to how the test plan is written, with an emphasis on making the traffic portion its own subtest?

```
t.Run("SINGLE_PRIMARY/persistence=DELETE"), func(t *testing.T) {
  clientA := ...
  t.Run("AddRoute", func(t *testing.T) {
    defer clientA.Close(t)
    ...
    t.Run("VerifyAFT", func(...) {...})
    t.Run("VerifyTraffic", func(...) {...})
  })
  t.Run("ShouldDelete", func(t *testing.T) {
    t.Run("VerifyNoAFT", func(...) {...})
    t.Run("VerifyNoTraffic", func(...) {...})
  })
})

t.Run("SINGLE_PRIMARY/persistence=PRESERVE"), func(t *testing.T) {
  clientA := ...
  t.Run("AddRoute", func(t *testing.T) {
    defer clientA.Close(t)
    ...
    t.Run("VerifyAFT", func(...) {...})
    t.Run("VerifyTraffic", func(...) {...})
  })
  t.Run("ShouldPreserve", func(t *testing.T) {
    t.Run("VerifyAFT", func(...) {...})
    t.Run("VerifyTraffic", func(...) {...})
  })
  t.Run("DeleteRoute", func(t *testing.T) {
    ...
    t.Run("VerifyNoAFT", func(...) {...})
    t.Run("VerifyNoTraffic", func(...) {...})
  })
})
```

It's recommended to use a symbolic name for the test name argument of `t.Run`. This will help future tooling that analyzes the test run results to identify the parts that failed. The go test output would say something like "FAIL: SINGLE_PRIMARY/persistence=DELETE/ShouldDelete/VerifyNoAFT" which is a succinct summary of the failure.

Reply:
Done. I have restructured the code with nested t.Run()s as recommended. Makes sense. Also reworded the help strings to be shorter and less verbose. Thanks!